### ISSUE TITLE
return __import__("pyodbc") ImportError: DLL load failed: The specified module could not be found.

### ISSUE THREAD

**Issue by Elthas17** [mkleehammer/pyodbc#663]
### Environment
- Python: Python 3.7.5  [MSC v.1916 64 bit (AMD64)] on win32
- pyodbc:  4.0.28
- OS: Microsoft Windows 10 Enterprise Version 10.0.17763 
- DB: Microsoft SQL Server Enterprise (64-bit) 14
- driver: ODBC Driver 17 for SQL Server

packages
- Flask==1.1.1
- Flask-Cors==3.0.8
- Flask-SQLAlchemy==2.4.1
- SQLAlchemy==1.3.12
- connexion==2.5.1

### Issue

Hey, thank you for pyodbc.

I upgraded from 4.0.27 (works fine)-> 4.0.28 and receive the following error / observed behaviour:

> Traceback (most recent call last):
>   File "[redacted]\venv\lib\site-packages\flask\app.py", line 2446, in wsgi_app
>     response = self.full_dispatch_request()
>   File "[redacted]\venv\lib\site-packages\flask\app.py", line 1951, in full_dispatch_request
>     rv = self.handle_user_exception(e)
>   File "[redacted]\venv\lib\site-packages\flask_cors\extension.py", line 161, in wrapped_function
>     return cors_after_request(app.make_response(f(*args, **kwargs)))
>   File "[redacted]\venv\lib\site-packages\flask\app.py", line 1820, in handle_user_exception
>     reraise(exc_type, exc_value, tb)
>   File "[redacted]\venv\lib\site-packages\flask\_compat.py", line 39, in reraise
>     raise value
>   File "[redacted]\venv\lib\site-packages\flask\app.py", line 1949, in full_dispatch_request
>     rv = self.dispatch_request()
>   File "[redacted]\venv\lib\site-packages\flask\app.py", line 1935, in dispatch_request
>     return self.view_functions[rule.endpoint](**req.view_args)
>   File "[redacted]\venv\lib\site-packages\connexion\decorators\decorator.py", line 48, in wrapper
>     response = function(request)
>   File "[redacted]\venv\lib\site-packages\connexion\decorators\uri_parsing.py", line 172, in wrapper
>     response = function(request)
>   File "[redacted]\venv\lib\site-packages\connexion\decorators\parameter.py", line 121, in wrapper
>     return function(**kwargs)
>   File "[redacted]\application\endpoints\redacted_my_python_file.py", line 238, in my_redacted_function
>     q_out = db.get_engine(None, 'redacted_SQL_ALCHEMY_BIND').execute(q).fetchall()
>   File "[redacted]\venv\lib\site-packages\flask_sqlalchemy\__init__.py", line 956, in get_engine
>     return connector.get_engine()
>   File "[redacted]\venv\lib\site-packages\flask_sqlalchemy\__init__.py", line 561, in get_engine
>     self._engine = rv = self._sa.create_engine(sa_url, options)
>   File "[redacted]\venv\lib\site-packages\flask_sqlalchemy\__init__.py", line 966, in create_engine
>     return sqlalchemy.create_engine(sa_url, **engine_opts)
>   File "[redacted]\venv\lib\site-packages\sqlalchemy\engine\__init__.py", line 479, in create_engine
>     return strategy.create(*args, **kwargs)
>   File "[redacted]\venv\lib\site-packages\sqlalchemy\engine\strategies.py", line 87, in create
>     dbapi = dialect_cls.dbapi(**dbapi_args)
>   File "[redacted]\venv\lib\site-packages\sqlalchemy\connectors\pyodbc.py", line 37, in dbapi
>     return __import__("pyodbc")
> ImportError: DLL load failed: The specified module could not be found.

I can replicate/remove the error by upgrading/downgrading pyodbc. 

expected behaviour:
no error

**Comment by pconwell** [comment]
I'm getting the same error (with python 3.8 - though I suspect that's not part of the problem).

**Comment by pconwell** [comment]
After some more digging - I'm not really sure what happened, but it seems to be related to `Build Tools for Visual Studio 2019` and the `C++ build tools`. Just to recap my chain of events:

1. Installed 4.0.28. No errors during install but wouldn't run (`DLL load failed` error on run).
2. Uninstalled 28 and tried to install 27. Got an error about C++ compiler during install. Tried 26, same error during install
3. Tried to reinstall 28, got C++ compiler install error again.
4. Installed `MSVC v142 - VS 2019 C++ x64/x86 build tools`, `C++ CMAKE tools for Windows`, and `c++ AddressSanitizer (Experimental)` from the `Build Tools for Visual Studio 2019`
5. Reinstalled 28

Now it seems to be working. I don't know why it would let me install 28 the first time without an error, but after installing updated C++ tools from build tools for VS it seems to have fixed itself.

**Comment by gordthompson** [comment]
@pconwell - @Elthas17 reported running 64-bit Python. Is that true for you as well? (I'm puzzled why the build tools would be involved. When I install 4.0.28 under 64-bit Python, pip just grabs the wheel file.)

**Comment by pconwell** [comment]
@gordthompson yes: `Python 3.8.0 (tags/v3.8.0:fa919fd, Oct 14 2019, 19:37:50) [MSC v.1916 64 bit (AMD64)] on win32`.

Like I said, I'm confused myself. The first time I installed 4.0.28 it installed fine. It wouldn't run so I tried to install 4.0.27 which failed during install complaining about build tools C++. Tried to reinstall 4.0.28 and it gave me the same build tools C++ error when trying to install (even though it had installed fine 2 minutes earlier).

Ended up installed C++ from the Visual Studio Build Tools and reinstalled 4.0.28 and it's working fine now.

???

**Comment by gordthompson** [comment]
I am able to reproduce the issue with a basic install of Python 3.7.6 64-bit:

```
C:\Users\Gord\AppData\Local\Programs\Python\Python37>python
Python 3.7.6 (tags/v3.7.6:43364a7ae0, Dec 19 2019, 00:42:30) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyodbc
>>> pyodbc.version
'4.0.27'
>>> ^Z


C:\Users\Gord\AppData\Local\Programs\Python\Python37>python -m pip install --upgrade pyodbc
Collecting pyodbc
  Downloading https://files.pythonhosted.org/packages/fa/b1/e7a3d321144e7701ad0850b82cb727407a273c80d5917a3acd2273ac6ec4/pyodbc-4.0.
28-cp37-cp37m-win_amd64.whl (66kB)
     |████████████████████████████████| 71kB 381kB/s
Installing collected packages: pyodbc
  Found existing installation: pyodbc 4.0.27
    Uninstalling pyodbc-4.0.27:
      Successfully uninstalled pyodbc-4.0.27
Successfully installed pyodbc-4.0.28

C:\Users\Gord\AppData\Local\Programs\Python\Python37>python
Python 3.7.6 (tags/v3.7.6:43364a7ae0, Dec 19 2019, 00:42:30) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyodbc
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: DLL load failed: The specified module could not be found.
>>>
```

The wheels at https://www.lfd.uci.edu/~gohlke/pythonlibs/#pyodbc have not yet been updated to 4.0.28, so they won't serve as a workaround for the latest version, but that site remains a way to get 4.0.27 for Python 3.8.

**Comment by CodeX1101** [comment]
> @gordthompson yes: `Python 3.8.0 (tags/v3.8.0:fa919fd, Oct 14 2019, 19:37:50) [MSC v.1916 64 bit (AMD64)] on win32`.
> 
> Like I said, I'm confused myself. The first time I installed 4.0.28 it installed fine. It wouldn't run so I tried to install 4.0.27 which failed during install complaining about build tools C++. Tried to reinstall 4.0.28 and it gave me the same build tools C++ error when trying to install (even though it had installed fine 2 minutes earlier).
> 
> Ended up installed C++ from the Visual Studio Build Tools and reinstalled 4.0.28 and it's working fine now.
> 
> ???

I experienced the same issue and also resolved by installing the latest vcredist for the 2015-2019 vc++

***note I already had that redist., but installing the latest version may have updated it

**Comment by gordthompson** [comment]
I can confirm the report from @CodeX1101 that installing the latest "Microsoft Visual C++ Redistributable for Visual Studio 2015, 2017 and 2019" from the links at

https://support.microsoft.com/en-us/help/2977003/the-latest-supported-visual-c-downloads

([vc_redist.x64.exe](https://aka.ms/vs/16/release/vc_redist.x64.exe) in my case) resolves the issue.

**Comment by cheenamalhotra** [comment]
It could be possible if the pyodbc **4.0.28** driver was built using later Visual C++ Redist packages than **4.0.27**, that this is noticeable.

@mkleehammer could you please confirm if Microsoft Visual C++ Redist package were upgraded on build machine for building **4.0.28** and if that differed from the state when **4.0.27** was built?

I'm not sure if that would need a revision or just documentation updates to ensure users install the latest "**Visual C++ Redistributable for Visual Studio 2015, 2017 and 2019**" in order to work with **4.0.28**.

**Comment by cbailiss** [comment]
Hi all
Just to confirm I am another person affected by this.  Thanks for the research on it @gordthompson 
Re: the below from @cheenamalhotra 
>> I'm not sure if that would need a revision or just documentation updates
I am not sure if this is a normal thing to do or not, but it seems like a potentially nasty gotchya for users who think they are just installing a small package update.

**Comment by cbailiss** [comment]
Investigating a little more, I found the following in my case:
MVC++ Redist. version installed originally:
Microsoft Visual C++ 2017 Redistributable
14.14.26429
Installing the latest version replaced the above with:
Microsoft Visual C++ 2015-2019 Redistributable
14.24.28127
This resolved the "import" error for me too.

As a further question - unless I have missed it somewhere, I didn't see any reference to the VC++ redistributable being a required pre-req for pyodbc in the pyodbc docs (e.g. in the installation section).  Is that because it is a general pre-req for Python on Windows?  I am asking because, if it is a specific pre-req for pyodbc, it would be helpful to note that in the installation section in pyodbc docs.

**Comment by ivarstange** [comment]
Hi, looks like you have done some investigations already. Here are my observations: I use `tox` to run my test suite against both Python 3.6 and 3.7. `pyodbc` is a package requirement and I only get the error message for the tests run against 3.7. The tests running with 3.6 are running smoothly. Pinning the `pyodbc` version to `4.0.27` makes it work for both Python versions.

**Comment by ivarstange** [comment]
Looks like you have done some investigation already. My observations are this: I'm using tox to run my test suite against Python 3.6 and Python 3.7. Version 4.0.28 of pyodbc works for Python 3.6 but not 3.7

**Comment by mathewrosssmith** [comment]
Just to add to this string, I had this same problem on a windows 10 machine after downloading python 3.8.1 and pyodbc 4.0.28 (this was a new desktop so neither of these were updates, just downloaded these versions initially). The problem was fixed with the "vs_redist.x64.exe 2019 64-bit" download at https://www.itechtics.com/microsoft-visual-c-redistributable-versions-direct-download-links/#Installation. Definitely seems like the Visual C++ is to blame (or the lack of requiring it).

**Comment by mhenke** [comment]
I came across this after upgrading to 4.0.27 from 4.0.28

![Capture](https://user-images.githubusercontent.com/119844/71843857-6bc64f00-308a-11ea-8312-7b8289875e73.PNG)

**Comment by mhenke** [comment]
When I try to go back to 4.0.27 after uninstalling I get an error also

![image](https://user-images.githubusercontent.com/119844/71843985-c6f84180-308a-11ea-8f56-be9f17da6130.png)

**Comment by andrsGutirrz** [comment]
Hello Guys, I solved this by downgrade pyodbc to 4.0.27
-------------------
Python version == 3.7

**Comment by copyrite** [comment]
I am also encountering the same problem with pyodbc 4.0.28 while with 4.0.27 everything works as intended.

**Comment by maycownd** [comment]
> 
> 
> Hello Guys, I solved this by downgrade pyodbc to 4.0.27
> 
> Python version == 3.7

It worked for me too.

**Comment by nayamp** [comment]
Downgrading to 4.0.27 on Python 3.7 also worked for me. Everything works just as it should now.

**Comment by j3ffer** [comment]
> Just to add to this string, I had this same problem on a windows 10 machine after downloading python 3.8.1 and pyodbc 4.0.28 (this was a new desktop so neither of these were updates, just downloaded these versions initially). The problem was fixed with the "vs_redist.x64.exe 2019 64-bit" download at https://www.itechtics.com/microsoft-visual-c-redistributable-versions-direct-download-links/#Installation. Definitely seems like the Visual C++ is to blame (or the lack of requiring it).

thanks, it's worked with me

**Comment by PNKonstant** [comment]
Downgrading to 4.0.27 on Python 3.7 also worked for me too!

**Comment by jkyeung** [comment]
> Re: the below from @cheenamalhotra
> 
> > I'm not sure if that would need a revision or just documentation updates
> 
> I am not sure if just documentation updates is a normal thing to do or not, but it seems like a potentially nasty gotchya for users who think they are just installing a small package update.

I'll be more blunt about it: A documentation update is not sufficient.

Most Python users probably do not check the documentation. The person updating the Python packages on a given system might not even be a programmer. They might just be doing "normal maintenance and upgrades".

Even if they look at the documentation, they might not have the knowledge, expertise, or inclination to mess with "redist" packages. Just witness all the respondents on this very thread who have decided their best course of action is to revert to the previous working version. That's even after someone posted a handy-dandy link with instructions and downloads. (Full disclosure: I went the downgrade route as well.)

**Comment by dhamacher** [comment]
> 
> 
> > Just to add to this string, I had this same problem on a windows 10 machine after downloading python 3.8.1 and pyodbc 4.0.28 (this was a new desktop so neither of these were updates, just downloaded these versions initially). The problem was fixed with the "vs_redist.x64.exe 2019 64-bit" download at https://www.itechtics.com/microsoft-visual-c-redistributable-versions-direct-download-links/#Installation. Definitely seems like the Visual C++ is to blame (or the lack of requiring it).
> 
> thanks, it's worked with me
Download link https://support.microsoft.com/en-ca/help/2977003/the-latest-supported-visual-c-downloads
Thanks, for me too.

**Comment by drrte5t** [comment]
Python 3.7 == pyodbc 4.0.27. It worked for me too!

**Comment by Jamspal** [comment]
python 3.8.1 and pyodbc 4.0.28 running on server2019
Download link https://support.microsoft.com/en-ca/help/2977003/the-latest-supported-visual-c-downloads

This worked for me, no need to downgrade

**Comment by mkleehammer** [comment]
I can guess what happened.  I started building pyodbc on a new Windows 10 laptop.  I had been building the official version for years on an ancient Windows 7 laptop with very old versions of Visual Studio.

I thought the MS universal CRT was supposed to fix all of this, but it looks like it is just going to shuffle the problem around.  Does anyone have any good docs for exactly how to fix this?

Any user of pyodbc is going to have Python installed.  Therefore any DLLs needed for Python are also installed.  I need pyodbc to link to the exact same version of the libs as the original Python build.

This is a mess.  I guess on Windows you can never install a newer version of Visual Studio than the one that was used to build the oldest version of Python you want to support?  That can't be right.

**Comment by mkleehammer** [comment]
I was able to reproduce with an older Windows 7 VM I had.  The problem is that 4.0.28 is linking against VCRUNTIME140_1.DLL.  Note that this is different than VCRUNTIME140.DLL which is included with Python.  (I was testing with 3.7.)

It is importing __CxxFrameHandler4 which was added in [Visual Studio 2019](http://www.nuonsoft.com/blog/tag/visual-studio-2019/ ).  I worked exclusively on Windows C++ projects for decades and I swear MS has never been able make their dev libraries just work.  It looks like compiling with 2019 will require a new DLL.  Perhaps there is a way to turn off the new exception handling for x64 and use the older, more bloated version?

As a temporary workaround, I'm going to rebuild on the VM and upload a new version.

**Comment by mkleehammer** [comment]
I found out how to turn off the new exception handling with some hidden command line options.   Fixed!  Thanks everyone.

**Comment by keitherskine** [comment]
Just wanted to add this URL to this issue, for reference:
https://devblogs.microsoft.com/cppblog/making-cpp-exception-handling-smaller-x64/

**Comment by rb2594** [comment]
> > Hello Guys, I solved this by downgrade pyodbc to 4.0.27
> > Python version == 3.7
> 
> It worked for me too.

Well, thanks it worked for me also.
I also got DLL load failed as error, so what I did is:
> `pip3 install --upgrade pyodbc`          # pyodbc upgraded from .28 to .30 (.30 is latest as of now)
Then, I ran my app on Anaconda prompt and it's running fine.

I think, if you have Visual C++ build tools installed then you don't have to mess around upgrading or downgrading pyodbc module.
